recipes/quickstart/agents/Agents_101/Tool_Calling_101.ipynb

```diff
@@ -0,0 +1,980 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Using this knowledge in practise\n",
+    "\n",
+    "A common misconception about tool calling is: the model can handle the tool call and get your output. \n",
+    "\n",
+    "This is NOT TRUE, the actual tool call is something that you have to implement. With this knowledge, let's see how we can utilise brave search to answer our original question"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 97,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#!pip3 install brave-search"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 98,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Assistant: <|python_tag|>wolfram_alpha.call(query=\"square root of 23131231\")\n"
+     ]
+    }
+   ],
+   "source": [
+    "SYSTEM_PROMPT = \"\"\"\n",
+    "Environment: iPython\n",
+    "Tools: brave_search, wolfram_alpha\n",
+    "Cutting Knowledge Date: December 2023\n",
+    "Today Date: 15 September 2024\n",
+    "\"\"\"\n",
+    "\n",
+    "user_input = \"\"\"\n",
+    "What is the square root of 23131231?\n",
+    "\"\"\"\n",
+    "\n",
+    "print(\"Assistant:\", model_chat(user_input, sys_prompt=SYSTEM_PROMPT))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 99,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "<|python_tag|>wolfram_alpha.call(query=\"square root of 23131231\")\n"
+     ]
+    }
+   ],
+   "source": [
+    "print(model_chat(user_input, sys_prompt=SYSTEM_PROMPT))\n",
+    "\n",
+    "output = model_chat(user_input, sys_prompt=SYSTEM_PROMPT)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 102,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Function name: wolfram_alpha\n",
+      "Method: call\n",
+      "Args: \"square root of 23131231\"\n"
+     ]
+    }
+   ],
+   "source": [
+    "import re\n",
+    "\n",
+    "# Extract the function name\n",
+    "fn_name = re.search(r'<\\|python_tag\\|>(\\w+)\\.', output).group(1)\n",
+    "\n",
+    "# Extract the method\n",
+    "fn_call_method = re.search(r'\\.(\\w+)\\(', string).group(1)\n",
+    "\n",
+    "# Extract the arguments\n",
+    "fn_call_args = re.search(r'=\\s*([^)]+)', string).group(1)\n",
+    "\n",
+    "print(f\"Function name: {fn_name}\")\n",
+    "print(f\"Method: {fn_call_method}\")\n",
+    "print(f\"Args: {fn_call_args}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "You can implement this in different ways but the idea is the same, the LLM gives an output with the `<|python_tag|>`, which should call a tool-calling mechanism. \n",
+    "\n",
+    "This logic gets handled in the program and then the output is passed back to the model to answer the user"
```

Comment:
@init27 it might be good if we could complete this part and show how to handle the actual tool calling from here